File: python/Q104_maximum_depth_of_binary_tree.py
#
# @lc app=leetcode id=104 lang=python3
#
# [104] Maximum Depth of Binary Tree
#


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None) -> None:
        self.val: int = val
        self.left: TreeNode | None = left
        self.right: TreeNode | None = right


# @lc code=start
class Solution:
    def maxDepth(self, root: TreeNode | None) -> int:
        if not root:
            return 0

        result: int = 0

        if (depth := self.maxDepth(root.left)) > result:
            result = depth

        if (depth := self.maxDepth(root.right)) > result:
            result = depth

        return result + 1


# @lc code=end

File: rust/src/leetcode/Q104_maximum_depth_of_binary_tree.rs
/*
 * @lc app=leetcode id=104 lang=rust
 *
 * [104] Maximum Depth of Binary Tree
 */

// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

// @lc code=start
use std::cell::RefCell;
use std::cmp::max;
use std::rc::Rc;

impl Solution {
    pub fn max_depth(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        if let Some(node) = root {
            let left_count = Self::max_depth(node.borrow_mut().left.clone());
            let right_count = Self::max_depth(node.borrow_mut().right.clone());

            max(left_count, right_count) + 1
        } else {
            0
        }
    }
}
// @lc code=end
